Sources/ComposableArchitecture/SwiftUI/ForEachStore.swift

```diff
@@ -1,6 +1,77 @@
 import OrderedCollections
 import SwiftUI
 
+/// A protocol that describes a collection of states with stable identifiers.
+///
+/// The library ships with two direct adopters of this protocol: `IdentifiedArray` from our
+/// [Identified Collections][swift-identified-collections] library, and `OrderedDictionary` from
+/// [Swift Collections][swift-collections].
+///
+/// [swift-identified-collections]: http://github.com/pointfreeco/swift-identified-collections
+/// [swift-collections]: http://github.com/apple/swift-collections
+public protocol IdentifiedStatesCollection: StateContainer {
+  typealias ID = Tag
+  associatedtype IDs: RandomAccessCollection where IDs.Element == ID
+  associatedtype States: Collection where States.Element == State
+
+  /// A `RandomAccessCollection` of ``ID`` which is in a 1-to-1 relationship with the ``states``
+  /// collection of ``StateContainer/State``.
+  ///
+  /// - Warning: You are responsible for keeping these values in a 1-to-1 relationship with
+  /// ``states``, which means that a ``StateContainer/State`` value should exist for any ``ID`` from
+  /// ``stateIDs``.
+  var stateIDs: IDs { get }
```

Comment:
There is no enforced requirement between `stateIDs` and `states`, but a the documentation warns that `stateIDs -> states` should be injective (I should probably rename 1-to-1 that can be ambiguous). We could probably require it to be bijective, but it would forbid us to define slices where the subset of `ID`s is smaller than `State`s.